crates/matrix-sdk/src/oidc/mod.rs

```diff
@@ -434,6 +437,157 @@ impl Oidc {
+            AuthorizationResponse::Error(err) => {
+                if err.error.error == AccessDenied {
+                    // The user cancelled the login in the web view.
+                    return Err(OidcError::CancelledAuthorization.into());
+                }
+                return Err(OidcError::Authorization(err).into());
+            }
+        };
+
+        // Given the InvalidState error already exists, I wonder if this was ever
+        // necessary to manually check?
+        if code.state != authorization_data.state {
+            return Err(OidcError::InvalidState.into());
+        };
+
+        self.finish_authorization(code).await?;
+        self.finish_login().await?;
+
+        Ok(())
+    }
+
+    /// Higher level wrapper that restores the OIDC client with automatic
+    /// static/dynamic client registration.
+    async fn configure(
+        &self,
+        issuer: String,
+        client_metadata: VerifiedClientMetadata,
+        registrations: OidcRegistrations,
+    ) -> std::result::Result<(), OidcError> {
+        if self.client_credentials().is_some() {
+            tracing::info!("OIDC is already configured.");
+            return Ok(());
+        };
+
+        if self.load_client_registration(issuer.clone(), client_metadata.clone(), &registrations) {
+            tracing::info!("OIDC configuration loaded from disk.");
+            return Ok(());
+        }
+
+        tracing::info!("Registering this client for OIDC.");
+        let registration_response =
+            self.register_client(&issuer, client_metadata.clone(), None).await?;
+
+        // The format of the credentials changes according to the client metadata that
+        // was sent. Public clients only get a client ID.
+        let credentials =
+            ClientCredentials::None { client_id: registration_response.client_id.clone() };
+        self.restore_registered_client(issuer, client_metadata, credentials);
+
+        tracing::info!("Persisting OIDC registration data.");
+        self.store_client_registration(&registrations)
+            .map_err(|e| OidcError::UnknownError(Box::new(e)))?;
+
+        Ok(())
+    }
+
+    /// Stores the current OIDC dynamic client registration so it can be re-used
+    /// if we ever log in via the same issuer again.
+    fn store_client_registration(
+        &self,
+        registrations: &OidcRegistrations,
+    ) -> std::result::Result<(), OidcError> {
+        let issuer = Url::parse(self.issuer().ok_or(OidcError::MissingAuthenticationIssuer)?)
+            .map_err(|e| OidcError::Url(e))?;
+        let client_id =
+            self.client_credentials().ok_or(OidcError::NotRegistered)?.client_id().to_owned();
+
+        registrations
+            .set_and_write_client_id(ClientId(client_id), issuer)
+            .map_err(|e| OidcError::UnknownError(Box::new(e)))?;
+
+        Ok(())
+    }
+
+    /// Attempts to load an existing OIDC dynamic client registration for a
+    /// given issuer.
```

Comment:
Please can you document the return value? (If there were a meaningful way to return `Result<(), SomeError>` that would be even better...

Reply:
Have added the documentation - it is perfectly valid to not find an existing registration so a `bool` seemed more appropriate to me than creating an error in this instance.

Reply:
Great, thanks!